### PR TITLE
Kunlun var - add functions to query block/time to ledger

### PR DESCRIPTION
**Pull request by leafyoung**

### CODE REVIEW

File: credmark/model/ledger/ledger.py [hunk 1]
```diff
@@ -144,3 +149,32 @@ def get_erc20_transfers(self, columns: List[str],
                                      TokenTransferTable,
                                      columns, where, group_by,
                                      order_by, limit, offset)
+
+    def get_block_number_of_date(self, dt_in: Union[datetime, date]):
```
Comment (iwyrkore) [inline credmark/model/ledger/ledger.py:153]: change name to: get_block_number_for_date

Use better name for parameters. `timestamp_date` ?

File: credmark/model/ledger/ledger.py [hunk 2]
```diff
@@ -144,3 +149,32 @@ def get_erc20_transfers(self, columns: List[str],
...
+        txn_blocks = self.Block.Columns
+        dt = datetime(dt_in.year, dt_in.month, dt_in.day, tzinfo=timezone.utc)
+        res = self.get_blocks(
```
Comment (iwyrkore) [inline credmark/model/ledger/ledger.py:156]: use model rpc.get_blocknumber

File: credmark/model/ledger/ledger.py [hunk 3]
```diff
@@ -144,3 +149,32 @@ def get_erc20_transfers(self, columns: List[str],
...
+            where=f"{txn_blocks.TIMESTAMP} >= {int(dt.timestamp())}",
+            order_by=f'{txn_blocks.TIMESTAMP} ASC',
+            limit='1')
+
+        rows = res.data
+        block_hist = rows[0].get(txn_blocks.NUMBER) if len(rows) else None
+
+        return block_hist
+
+    def get_date_of_block_number(self, block_hist: int):
```
Comment (iwyrkore) [inline credmark/model/ledger/ledger.py:167]: change `block_hist` to `block_number`

Comment (iwyrkore) [reply credmark/model/ledger/ledger.py:167]: change to  `get_date_for_block_number`

File: credmark/model/ledger/ledger.py [hunk 4]
```diff
@@ -144,3 +149,32 @@ def get_erc20_transfers(self, columns: List[str],
...
+        txn_blocks = self.Block.Columns
+        res = self.get_blocks(
```
Comment (iwyrkore) [inline credmark/model/ledger/ledger.py:169]: use model rpc.get-block-timestamp in model-runner which returns {"timestamp": 123}

**Comment by iwyrkore** [comment]
@leafyoung In fact, you can already get the timestamp for a block number by doing:

```
b = BlockNumber(blockNumber)
b.to_datetime()
```

ALTHOUGH that `to_datetime()` looks broken because it doesn't use the UTC datetime so you can fix that.

For getting a block number for a timestamp do:

```
b = BlockNumber(sample_timestamp=some_timestamp)
```

**Comment by leafyoung** [comment]
In general, these helpers are to obtain the block number just before or at certain point of time, and vice versa. 
- get_block_number from a datetime? This needs the ledger
- get datetime from a block_number? This does not need ledger and I can improve on the current version.

**Comment by iwyrkore** [comment]
The BlockNumber class is using the ledger (and web3). See my comment above.

File: credmark/model/ledger/ledger.py (hunk 1 above)
Comment (leafyoung) [reply credmark/model/ledger/ledger.py:153]: renamed

**Comment by leafyoung** [comment]
> @leafyoung The BlockNumber class is using the ledger (and web3). See my comment above.
> 
> It currently uses `rpc.get-blocknumber` to get the block number from a timestamp, passed as `sample_timestamp` (I'm not sure we need to make the type a web3 Timestamp though... that seems limiting. Could be int or Timestamp or None.
> 
> ```
> b = BlockNumber(sample_timestamp=Timestamp(1454544333))
> ```

query by sample_timestamp only is limiting.

File: credmark/model/ledger/ledger.py (hunk 2 above)
Comment (leafyoung) [reply credmark/model/ledger/ledger.py:156]: moved to block_number and use rpc.get_blocknumber for date -> block_number.

File: credmark/types/data/block_number.py [hunk 5]
```diff
@@ -100,15 +113,52 @@ def __timestamp__(self) -> Timestamp:
...
+            dt = datetime.fromtimestamp(int(datetime_date_ts), tz=timezone.utc)
+        elif isinstance(datetime_date_ts, datetime):
+            dt = datetime_date_ts.replace(tzinfo=timezone.utc)
+        elif isinstance(datetime_date_ts, date):
+            dt = datetime.combine(datetime_date_ts, datetime.max.time(), tzinfo=timezone.utc)
+        else:
+            raise ModelInputError(f'Invalid input for date/datetime/timestamp to query block_number {datetime_date_ts=}')
+        ts = int(dt.timestamp())
+
+        block_number = context.models.rpc.get_blocknumber(input=dict(timestamp=ts))['blockNumber']
```
Comment (iwyrkore) [inline credmark/types/data/block_number.py:150]: Just `return cls(sample_timestamp=ts)` You don't need to call the rpc, it's already being done in the `__new__` method.

However, it seems the `__new__` method is missing the max_block_number check when it looks up by timestamp. It should raise `BlockNumberOutOfRangeError` so you can fix that.

File: credmark/model/ledger/ledger.py [hunk 6]
```diff
@@ -1,7 +1,20 @@
+from datetime import (
```
Comment (iwyrkore) [inline credmark/model/ledger/ledger.py:1]: I assume these imports aren't needed anymore

File: credmark/model/ledger/ledger.py [hunk 7]
```diff
@@ -1,7 +1,20 @@
...
-from .errors import InvalidColumnException, InvalidQueryException
+from .errors import (
+    InvalidColumnException,
+    InvalidQueryException,
+)
+
+from credmark.model.errors import (
+    ModelInputError,
+)
+
```
Comment (iwyrkore) [inline credmark/model/ledger/ledger.py:17]: I assume these imports aren't needed anymore

**Comment by iwyrkore** [comment]
> query by sample_timestamp only is limiting.

Not sure what that means.

File: credmark/types/data/block_number.py [hunk 8]
```diff
@@ -100,15 +113,52 @@ def __timestamp__(self) -> Timestamp:
...
+        if block_number > context.block_number:
+            context.logger.warn(f'block_number {block_number} for {dt} is later than the current block {context.block_number} on {context.block_number.to_datetime()}. Return current block')
+            return cls(context.block_number)
+        else:
+            return cls(block_number)
+
+    @staticmethod
+    def datetime_of(block_number: int):
+        context = credmark.model.ModelContext.current_context
+        breakpoint()
```
Comment (iwyrkore) [inline credmark/types/data/block_number.py:160]: remove

File: credmark/types/data/block_number.py [hunk 9]
```diff
@@ -100,15 +113,52 @@ def __timestamp__(self) -> Timestamp:
...
+        timestamp = context.models.rpc.get_block_timestamp(input=dict(blockNumber=block_number))['timestamp']
+        breakpoint()
```
Comment (iwyrkore) [inline credmark/types/data/block_number.py:162]: remove

File: credmark/types/data/block_number.py [hunk 10]
```diff
@@ -100,15 +113,52 @@ def __timestamp__(self) -> Timestamp:
...
+        block_number = context.models.rpc.get_blocknumber(input=dict(timestamp=ts))['blockNumber']
+        if block_number > context.block_number:
+            context.logger.warn(f'block_number {block_number} for {dt} is later than the current block {context.block_number} on {context.block_number.to_datetime()}. Return current block')
+            return cls(context.block_number)
+        else:
+            return cls(block_number)
+
+    @staticmethod
+    def datetime_of(block_number: int):
+        context = credmark.model.ModelContext.current_context
```
Comment (iwyrkore) [inline credmark/types/data/block_number.py:159]: BlockNumber already has `to_datetime()` that should be fixed to return a utc timestamp. 
Do we need this static method as well?

**Comment by leafyoung** [comment]
@iwyrkore, thanks for all the points above and make rpc.get-block-timestamp available. I took a few thoughts and experiments. There are existing tools to query block number / block timestamp. However, the Python date and datetime has some details with timezone/timestamp. And there are some conventions for the end of date. I still feel some need for some utility to convert/query between the three, block number, block timestamp and Python datetime. Here is a write-up that I prepare to README. Hope it explains.

__Block number, Timestamp and Python datetime__

In blockchain, every block is created with a timestamp (in Unix epoch). While in Python there are two types for date, date and datetime, with datetime can further be set with tzinfo or without tzinfo. To provide convenient way to convert/query between the three and avoid some of the pitfalls in Python datetime, we have implemented a few utilities within the `BlockNumber` class. Below are their descriptions and examples.

1. instance method, `block_number.to_datetime(self)`:
    Return the Python datetime with UTC of the block

2. instance method, `block_number.timestamp(self)`:
    Return the Unix epoch of the block.

3. class method: `from_datetime(cls, datetime_date_ts: Union[datetime, date, int, float])`
    Obtain a BlockNumber to be less or equal to the input datetime, date (as of the end time of the day), and timestamp.

    Be caution when we obtain a timestamp from a Python datetime, we should attach a tzinfo (e.g. timezone.utc) to the datetime. Otherwise, Python take account of the local timezone when converting to a timestamp. See examples below.

4. static method: `datetime_of(block_number: int)`
    Obtain a Python datetime with UTC of the block

More example code for the block-number class can be found [1](https://github.com/credmark/credmark-model-framework-py/blob/main/credmark/types/data/block_number.py) and [2](https://github.com/credmark/credmark-models-py/blob/main/models/examples/block_time_example.py).

Note: The link to the example code is currently in
https://github.com/credmark/credmark-models-py/blob/kunlun-var/models/examples/block_time_example.py

**Comment by iwyrkore** [comment]
This looks good. I have a few suggestions:

> 
> 1. instance method, `block_number.to_datetime(self)`:
>    Return the Python datetime with UTC of the block.

Document it without `self` and I think a better name is `block_number.timestamp_datetime()` so it will show up next to `block_number.timestamp` in the docs and match the db field etc.

> 2. instance method, `block_number.timestamp(self)`:
>    Return the Unix epoch of the block.

Document it without `self` as you'd call it: `block_number.timestamp()`

> 3. class method: `from_datetime(cls, datetime_date_ts: Union[datetime, date, int, float])`
>    Obtain a BlockNumber to be less or equal to the input of datetime with tzinfo,  as of the end time of a day,  and a timestamp.

Instead we could change the `__new__` and `__init__` to accept sample_timestamp as `Union[datetime, date, int, float]`?  However, using a class method might be more obvious to people as a beginner might not understand the concept of `sample_timestamp`. 

> 4. static method: `datetime_of(block_number: int)`
>    Obtain a Python datetime with UTC of the block

I guess this is the case where you just have an int and don't want to convert it to a BlockNumber instance?

**Comment by leafyoung** [comment]
Thanks for the reply.
For 1 and 2, I will edit the doc as suggested.
For 3, a class method is easier to be picked up by the beginners.
For 4, yes. Some case, the user just want to do simple query for the time, not in need of creating an object.

File: credmark/types/data/block_number.py [hunk 11]
```diff
@@ -140,16 +140,21 @@ def from_datetime(cls, datetime_date_ts: Union[datetime, date, int, float]):
         elif isinstance(datetime_date_ts, float):
             dt = datetime.fromtimestamp(int(datetime_date_ts), tz=timezone.utc)
         elif isinstance(datetime_date_ts, datetime):
-            dt = datetime_date_ts.replace(tzinfo=timezone.utc)
+            if not datetime_date_ts.tzinfo:
+                raise ModelInputError(
+                    f'Input datetime {datetime_date_ts} has no tzinfo, please add.')
+            ts = int(datetime_date_ts.timestamp())
         elif isinstance(datetime_date_ts, date):
```
Comment (iwyrkore) [inline credmark/types/data/block_number.py:155]: I fixed this. It should be assigning datetime_date_ts to dt and then is gets the ts from it below.

**Pull request merged**